opm/material/fluidsystems/GenericFluidSystem.hh

```diff
@@ -0,0 +1,223 @@
+  module for the precise wording of the license and the list of
+  copyright holders.
+*/
+
+#ifndef OPM_GENERICFLUIDSYSTEM_HH
+#define OPM_GENERICFLUIDSYSTEM_HH
+
+#include <opm/common/OpmLog/OpmLog.hpp>
+
+#include <opm/material/fluidsystems/BaseFluidSystem.hpp>
+#include <opm/material/fluidsystems/PTFlashParameterCache.hpp> // TODO: this is something else need to check
+#include <opm/material/viscositymodels/LBC.hpp>
+
+#include <cassert>
+#include <string>
+
+#include <fmt/format.h>
+
+namespace Opm {
+
+    template <typename Scalar>
+    struct ComponentParam {
+        std::string name;
+        Scalar molar_mass;
+        Scalar critic_temp;
+        Scalar critic_pres;
+        Scalar critic_vol;
+        Scalar acentric_factor;
+
+        ComponentParam(const std::string_view name_, const Scalar molar_mass_, const Scalar critic_temp_,
+                       const Scalar critic_pres_, const Scalar critic_vol_, const Scalar acentric_factor_)
+                       : name(name_),
+                         molar_mass(molar_mass_),
+                         critic_temp(critic_temp_),
+                         critic_pres(critic_pres_),
+                         critic_vol(critic_vol_),
+                         acentric_factor(acentric_factor_)
+        {}
+    };
+/*!
+ * \ingroup FluidSystem
+ *
+ * \brief A two phase system that can contain NumComp components
+ *
+ * \tparam Scalar  The floating-point type that specifies the precision of the numerical operations.
+ * \tparam NumComp The number of the components in the fluid system.
+ */
+    template<class Scalar, int NumComp>
+    class GenericFluidSystem : public BaseFluidSystem<Scalar, GenericFluidSystem<Scalar, NumComp> > {
+    public:
+        // TODO: I do not think these should be constant in fluidsystem, will try to make it non-constant later
+        static const int numPhases=2;
+        static const int numComponents = NumComp;
+        static const int numMisciblePhases=2;
+        static const int numMiscibleComponents = 3;
+        // TODO: phase location should be more general
+        static constexpr int oilPhaseIdx = 0;
+        static constexpr int gasPhaseIdx = 1;
+
+        template <class ValueType>
+        using ParameterCache = Opm::PTFlashParameterCache<ValueType, GenericFluidSystem<Scalar, NumComp>>;
+        using ViscosityModel = Opm::ViscosityModels<Scalar, GenericFluidSystem<Scalar, NumComp>>;
+        using PengRobinsonMixture = Opm::PengRobinsonMixture<Scalar, GenericFluidSystem<Scalar, NumComp>>;
+
+        template<typename Param>
+        static void addComponent(const Param& param)
+        {
+            assert(component_param_.size() <= numComponents);
+            if (component_param_.size() == numComponents) {
+                const std::string msg =  fmt::format("the fluid system has reached maximum {} component, the component {} will not be added",
+                                                     NumComp, param.name);
```

Comment:
It must be compiler related, if I use numComponents instead of NumComp here, I got the following link error.

```
/usr/bin/ld: CMakeFiles/co2_ptflash_ecfv.dir/tests/co2_ptflash_ecfv.cc.o: warning: relocation against `_ZN3Opm18GenericFluidSystemIdLi3EE13numComponentsE' in read-only section `.text._ZN3Opm18GenericFluidSystemIdLi3EE12addComponentINS_14ComponentParamIdEEEEvRKT_[_ZN3Opm18GenericFluidSystemIdLi3EE12addComponentINS_14ComponentParamIdEEEEvRKT_]'
/usr/bin/ld: CMakeFiles/co2_ptflash_ecfv.dir/tests/co2_ptflash_ecfv.cc.o: in function `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > fmt::v8::format<int const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&>(fmt::v8::basic_format_string<char, fmt::v8::type_identity<int const&>::type, fmt::v8::type_identity<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&>::type>, int const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/usr/include/fmt/core.h:3119: undefined reference to `Opm::GenericFluidSystem<double, 3>::numComponents'
/usr/bin/ld: warning: creating DT_TEXTREL in a PIE
collect2: error: ld returned 1 exit status
```